Fixes for eLibrary.ru

i3v opened the pull request:
Quick fixes for `eLibrary.ru.js` - it looks like they changed web page structure a bit, thus old translator fails to get list of authors and `itemType`.

- With this fix, tests 2 and 3 now pass OK in Zotero 5.0.68
- The first test `multiple` still fails. I'm not sure if it's possible to fix it.

zuphilip commented on eLibrary.ru.js:
```diff
@@ -291,7 +293,7 @@ var testCases = [
 				"date": "2012",
 				"ISSN": "2219-5335",
-				"issue": "13",
+				"issue": "13 (1)",
```
Is that volume and issue? Then we should try to split this up into two fields, whenever we see such a pattern.

However, this could also simply be the "issue number" relative to this year, i.e. it is overall the issue 13 but in the year 2012 it is issue 1.

i3v replied:
[This](https://elibrary.ru/contents.asp?id=33729006&selid=17339044) looks like this is the "issue number relative to this year":
![image](https://user-images.githubusercontent.com/920364/60020277-64b09400-9698-11e9-8a35-0b5fd43deef9.png)

However... [Here](https://elibrary.ru/contents.asp?titleid=10023)'s another journal, which uses similar format, but now "issue number relative to this year" comes first:
![image](https://user-images.githubusercontent.com/920364/60020387-99245000-9698-11e9-8b31-6b37f23290c9.png)

Thus, I'm not sure if it's safe to make any assumptions about the contents of that field.

zuphilip replied:
Okay, then we can leave it as it comes from the website, which then maybe is be the best we can do in a general approach.

i3v commented on eLibrary.ru.js:
```diff
@@ -105,8 +105,8 @@ function scrape (doc) {
-		var title = ZU.xpathText(datablock, "./table[1]");
-		var authors =   ZU.xpath(datablock, './div[1]/table[1]//span/a[contains(@href, "authorid")]');
+		
+		var authors = ZU.xpath(doc, '//table[@width=550]//td[@width=514]/span[@style="white-space: nowrap"]//b');
```
The `author_items.asp?authorid` won't work for the [2nd test case](https://elibrary.ru/item.asp?id=9541154) - author name is plain text there, not url.

The `<input type="hidden" name="authors_all" value=...` looks more structured. (Used for "find other papers by same authors" button), but its value is quite strange (see [3rd test case](https://elibrary.ru/item.asp?id=17339044)): 
```
<input type="hidden" name="authors_all" value="Седов Е|Седов Е Н|Седышева Г|Седышева Г А|Серова З|Серова З М|Супрун И|Супрун И И|Ульяновская Е В|Уляновская Е|">
```
 * names are shortened `Седов Е Н` instead of `СЕДОВ ЕВГЕНИЙ НИКОЛАЕВИЧ`
 * names are duplicated `Седов Е|Седов Е Н`
 * names order is incorrect (correct is `СУПРУН ИВАН ИВАНОВИЧ, УЛЬЯНОВСКАЯ ЕЛЕНА ВЛАДИМИРОВНА, СЕДОВ ЕВГЕНИЙ НИКОЛАЕВИЧ, ...`)

thus, this field is hardly useful.

I do agree that xpath, mainly based on `td[@width=514]` doesn't look nice. But, probably, anything else would be just as fragile as this.

I'll try to contact site admins, maybe they'll be so kind to adjust something...

zuphilip replied:
Okay, I understand. Thank you for looking this more up in the examples. We can also go here for the path you currently are suggesting and hope that maybe the publisher will change to something more structured website.

i3v commented on eLibrary.ru.js:
```diff
@@ -140,18 +140,20 @@ function scrape (doc) {
 			"Место издания" : "place",
 			"Цит. в РИНЦ" : "extra",
-			"Тип " : "itemType"
+			"Тип:" : "itemType"
```
Hm... I'm not sure I understand your idea. I've removed `:` from those lookup patterns in 7c6604c. But I'm completely missing the "make it cleaner" idea. Could you please point me into a proper, cleaner piece of code?

zuphilip replied:
I meant that you then can add the semi-colon `:` in this line
https://github.com/zotero/translators/blob/7c6604c3a03575a6be817a998c85867c7209d25f/eLibrary.ru.js#L170
i.e.
```js
		var t = ZU.xpathText(doc, '//tr/td/text()[contains(., "' + key + ':")]/following-sibling::*[1]');
```
Sorry, I haven't looked close enough to the code.

Does that work and are the tests still succeeding?

i3v replied:
OK, thanks for explaining. I'll do that. 
The only thing that gets broken is broken either way. 
I'll push few more commits in a couple of days, to support more item types.

zuphilip replied:
Okay, great. Let us know when you are through by just writing a new message here at the bottom.

i3v commented:
OK, I think that's it for now.
I've added few more fields (e.g. DOI), fixed some broken fields (e.g. abstract), improved test coverage.
All tests work OK now, including the first one (about multiple items).

zuphilip commented:
Thank you @i3v ! I did some final cleanup and squashed the commits together. (Note, usually it is not recommended to use the master branch for PR.)

i3v commented:
@zuphilip , thanks for merging and for the fixes!
I'll try to make it cleaner next time :)
